### PR TITLE
Add `UserProfilesStore`, `LruCache` and cache for user permalink profiles

### PR DESCRIPTION
**Pull request by weeman1337**
Adds the `UserProfilesStore` to cache global user profiles (global = retrieved via the profiles API).

**Review guide:**
-  Best to start with`LruCache`
- Then `UserProfileStore` that utilises `LruCache`
- Finally the changes of the `userPermalinkMember` hook

| Before | After |
| --- | --- |
| user_profile_before | user_profile_after |

part of https://github.com/vector-im/element-meta/issues/1030

closes https://github.com/vector-im/element-web/issues/10559

PSF-1961

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Notes: Added `UserProfilesStore`, `LruCache` and user permalink profile caching

element-web notes: User permalinks are now cached and therefore do less often flicker after being lazy loaded.

---
Here's what your changelog entry will look like:

## ✨ Features
 * Added `UserProfilesStore`, `LruCache` and user permalink profile caching ([\#10425](https://github.com/matrix-org/matrix-react-sdk/pull/10425)). Fixes vector-im/element-web#10559.

### CODE REVIEW

File: src/stores/UserProfilesStore.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+/**
+ * This store provides cached access to user profiles.
+ * Listens for membership events and invalidates the cache for a profile on update with different profile values.
+ */
+export class UserProfilesStore {
+    private profiles = new LruCache<string, IMatrixProfile | null>(cacheSize);
+    private knownProfiles = new LruCache<string, IMatrixProfile | null>(cacheSize);
+
+    public constructor(private client?: MatrixClient) {
```
Comment (t3chguy) [inline src/stores/UserProfilesStore.ts:34]: Why is the client optional?

Comment (weeman1337) [reply src/stores/UserProfilesStore.ts:34]: The client is optional in SDKContext.

Comment (t3chguy) [reply src/stores/UserProfilesStore.ts:34]: Sure, but what sense does constructing a UserProfileStore without a client make? Especially given it is a singleton. Without a client it'll never cull stale data. Along with the race condition listed below that sounds likely to introduce bugs.

File: src/contexts/SDKContext.ts [hunk 2]
```diff
@@ -185,4 +187,11 @@ export class SdkContextClass {
         }
         return this._AccountPasswordStore;
     }
+
+    public get userProfilesStore(): UserProfilesStore {
+        if (!this._UserProfilesStore) {
+            this._UserProfilesStore = new UserProfilesStore(this.client);
```
[the remark this answers is not in the document]
Comment (weeman1337) [reply src/contexts/SDKContext.ts:193]: Not sure what to do about that :frowning: . Client is optional in SDKContext. How about clearing the store after login and setting the client [here](https://github.com/matrix-org/matrix-react-sdk/blob/2e8e4747920694b07720e219341a7591b59090de/src/components/structures/MatrixChat.tsx#L800)?

[the remark this answers is not in the document]
Comment (t3chguy) [reply src/contexts/SDKContext.ts:193]: Yeah I think it needs to allow setting a matrixClient at a later stage (once initialised) and prune data and set up listeners when that happens

[the remark this answers is not in the document]
Comment (weeman1337) [reply src/contexts/SDKContext.ts:193]: None of this really makes sense. Without a client, you can't fetch profiles anyway. I tend to make the store client non-optional and raise an error here.

**Pull request merged**